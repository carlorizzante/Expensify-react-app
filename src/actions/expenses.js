import uuid from 'uuid';
import database from '../firebase/firebase.js';

// ADD_EXPENSE
export const addExpense = (expense) => {
  return {
    type: 'ADD_EXPENSE',
    expense
  }
}

// Connect Firebase, store data, dispatch "addExpense"
export const startAddExpense = (expense) => {
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = expense;

  return (dispatch) => {
    const expense = { description, note, amount, createdAt };
    return database.ref('expenses')
      .push(expense)
      .then((ref) => dispatch(addExpense({
        id: ref.key,
        ...expense
      })));
  }
}

// REMOVE_EXPENSE
export const removeExpense = id => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  }
}
// EDIT_EXPENSE
export const editExpense = (id, updates = {}) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updates
  }
}
