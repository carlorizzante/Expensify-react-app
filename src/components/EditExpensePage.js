import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm.js';
import {
  // editExpense,
  startEditExpense,
  // removeExpense,
  startRemoveExpense
} from '../actions/expenses.js';

export class EditExpensePage extends Component {
  onSubmit = (expense) => {
    // this.props.editExpense(this.props.expense.id, expense);
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  }
  onRemove = (id = this.props.expense.id) => {
    // this.props.removeExpense(id);
    this.props.startRemoveExpense(id);
    this.props.history.push("/");
  }
  render() {
    return (
      <div>
        <h1>Edit Expense</h1>
        <ExpenseForm
          expense={ this.props.expense }
          onSubmit={ this.onSubmit }
          onRemove={ this.onRemove }
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  const { id } = props.match.params;
  return {
    expense: state.expenses.find(expense => expense.id == id)
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  // editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  // removeExpense: (id) => dispatch(removeExpense(id)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
} );

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
