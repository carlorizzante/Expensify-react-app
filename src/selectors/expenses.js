import moment from 'moment';

// January 1st 1970 -> Unix Epoch
// timestamp in milliseconds

// Get Visible Espenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter(expense => {
    // const createdAt = expense.createdAt;
    const createdAtMoment = moment(expense.createdAt);
    // const startDateMatch = typeof startDate !== 'number' || createdAt >= startDate;
    // const endDateMatch = typeof endDate !== 'number' || createdAt <= endDate;
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
  }).sort((exp1, exp2) => {
    if (sortBy == 'amount') return exp1.amount < exp2.amount;
    else return exp1.createdAt < exp2.createdAt;
  });
}
