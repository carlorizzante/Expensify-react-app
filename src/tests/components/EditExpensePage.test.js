import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage.js';
import expenses from '../fixtures/expenses.js';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage
    expense={ expenses[2] }
    editExpense={ editExpense }
    removeExpense= { removeExpense }
    history={ history }
  />);
});

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('Connect(ExpenseForm)').prop('onSubmit')(expenses[2]);
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle onRemove', () => {
  wrapper.find('Connect(ExpenseForm)').prop('onRemove')(expenses[2].id);
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[2].id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});
