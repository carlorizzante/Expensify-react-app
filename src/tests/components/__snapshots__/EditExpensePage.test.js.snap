// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <Connect(ExpenseForm)
    expense={
      Object {
        "amount": 300,
        "createdAt": 864000000,
        "description": "Much later",
        "id": "vf45",
        "note": "Maecenas sed diam eget risus varius blandit sit amet non magna.",
      }
    }
    onRemove={[Function]}
    onSubmit={[Function]}
  />
</div>
`;
